### ISSUE TITLE
Add 'skipNavigation' boolean to routerLink to disable navigating 

### ISSUE THREAD

**Issue by TimGeerts** [angular/angular#31154]
# 🚀 feature request

### Relevant Package
This feature request is for `@angular/router` (specifically the `routerLink` directive)

### Description
At the moment it doesn't seem possible to combine the power of `routerLinkActive` (for class changes) with a `routerLink` that doesn't navigate.  
This can be useful for instance when you have a navigation of multiple levels, and only the child levels are allowed to actually navigate (but you still want to leverage the `routerLinkActive` directive to add/remove classes on the links that shouldn't navigate.

### Describe the solution you'd like
* Add `skipNavigation` to the `routerLink` directive to block the actual navigation of the `onClick` handler.

``` typescript
  @HostListener('click', ['$event.button', '$event.ctrlKey', '$event.metaKey', '$event.shiftKey'])
  onClick(button: number, ctrlKey: boolean, metaKey: boolean, shiftKey: boolean): boolean {
    if (button !== 0 || ctrlKey || metaKey || shiftKey) {
      return true;
    }

    if (typeof this.target === 'string' && this.target !== '_self') {
      return true;
    }

    const extras = {
      skipLocationChange: attrBoolValue(this.skipLocationChange),
      replaceUrl: attrBoolValue(this.replaceUrl),
      state: this.state
    };
+    if(!attrBoolValue(this.skipNavigation)) {
+      this.router.navigateByUrl(this.urlTree, extras);
+    }
    return false;
  }
```

### Describe alternatives you've considered
* making the `routerLink` of the non-navigating links empty (or /) doesn't help, since `routerLinkActive` will match all of them when using `{ exact: false }`.
* using `{exact: true}` works for not matching all the root levels, but this will in turn never actually show a clear view of what links are currently active based on the URL (current link + all it's parents should be set to active, not only the clicked link).
* good practice would be using a guard, but for this specific situation it seems to be quite the overkill when it can just be fixed with a boolean that is false by default (so doesn't impact anything but the actual case).

### Related Issues
#10431
#22588
#13980

**Comment by Airblader** [comment]
You could also create your own directive that applies a class when a route is matched, but doesn't do any kind of navigation. 

A other option is to apply routerLink on a disabled button as here the browser already takes care of not calling the click handler.

The boolean looks a bit hacky to me. The primary function of the routerLink directive is to navigate, not to add active classes. If you don't want it to be a link, it probably shouldn't be a link, IMHO.

**Comment by TimGeerts** [comment]
I agree that the primary function of the routerLink directive is navigation. And in general I'd also agree that "if you don't want it to be a link, it probably shouldn't be a link".
However, my current usecase is a side-navigation component, and I want to be able to allow for both interactions (root levels navigate or don't navigate) based on a setting they can use for my component.

Making my own directive solely for the purpose of matching classes is a possibility, yet I don't think adding that boolean would be considered a "hacky" approach since there's already some others in there as well that don't really handle navigation but handle url writes or location changes.

TLDR: I agree that there's probably other options (custom directive for styling) and that it might not be the core business of the routerLink directive, yet I feel it's a quick win without downsides if it's added to the actual directive. (opinionated, of course :) )

**Comment by Airblader** [comment]
> I feel it's a quick win without downsides

Also opinionated: a quick win still has to be maintained for the entire future of the project as otherwise it's a breaking change, and every feature comes with downsides, namely its complexity and contribution to bloat. The question is whether the benefit outweighs the cost. 

The other question to consider is not just how simple the change is, but whether it's a good/intuitive API.

I don't want to drag this out too much — simply gave my opinion above, it's of course up to the Angular team and given that this falls into the small feature category, there's a chance it would just be merged if you open a PR for it. :)

**Comment by jasonaden** [comment]
Thanks for the issue and good feedback @Airblader. I'm not sure this is the solution we would go with, but I get the concept. Maybe something around being able to stop the event of the click similar to how you can do with regular `a href`'s?

**Comment by TimGeerts** [comment]
I understand that you would do it differently, and yes, the feedback is very much appreciated (this was the first time that I ventured into "feature requests" for a big framework).  
For now, I've made my own directives (basically a copy of the routerlink and routerlinkactive directives) with this boolean added to it.  
Just thought I'd try to propose it as a feature request as I'd obviously prefer to use the default directives.
I'll see where this goes and will keep track of it.
Thanks all!

**Comment by keatkeat87** [comment]
![dada](https://user-images.githubusercontent.com/8691013/69030868-b117ba00-0a13-11ea-8b6f-b2894ce6408f.gif)

This is an user experience on google ads 

sidenav (for parent) + tabnav (for child)

sidenav routerLink = "/aw/keywords"
tab1 routerLink = ""
tab2 routerLink = "negative"

the user behavior is like that 
1. click sidenav (visit tab1)
2. click tab negative (visit tab2)
3. click sidenav again 
boom, because angular routerLink can't be stop, so the user will back to the tab1.
but in google ads, it won't back to the tab1.

i think this is a normal use case, maybe you can consider it.

**Comment by simeyla** [comment]
@jasonaden When a question is viewed on Stackoverflow [over 63k times](https://stackoverflow.com/questions/35431188/how-can-i-conditionally-disable-the-routerlink-attribute) shouldn't that give a request like this a bump?

Cancelling a link is something very fundamental to how people have learnt to make websites, and `CanActivate` / `CanDeactivate` are often too complex when you just need to stop the default action (for many reasons).

Something simple like `(routerLinkClick)` with parameters signature `(urlTree, extras)` would allow users to override the default `onClick()` handler and save thousands of hours looking for a solution. I've spent two hours today and probably another two hours a year ago on this exploring solutions.

I just can't bring myself to clone the whole of the [routerLink](https://github.com/angular/angular/blob/bb290cefae154fc7a2b0903b8371da4eb4adebc0/packages/router/src/directives/router_link.ts#L269) and [routerLinkActive](https://github.com/angular/angular/blob/master/packages/router/src/directives/router_link_active.ts#L6) files for something so trivial.

I really hope someday this can be added. Thanks.

**Comment by kfwerf** [comment]
think we had some issues with this as well, is this an issue we want to solve? im thinking i could pick it up and add some attribute that allows for a routerDisabled state while still being active, e.g. highlighted

**Comment by wchid** [comment]
While a built-in solution is in the works (https://github.com/angular/angular/pull/35393), I've had success preventing unwanted navigation when invoking navigation through `routerLink` children of a container wielding a non-navigational `routerLink` attribute (for `routerLinkActive` functionality). Apply the following directive to the container with the non-navigational `routerLink` attribute:
```ts
import {Directive, ElementRef, Input, OnDestroy, Renderer2} from "@angular/core";

@Directive({
	selector: "[skipNavigation][routerLink]",
})
export class SkipNavigationDirective implements OnDestroy {

	@Input() public skipNavigation = false;

	private unsubscribe: () => void;

	constructor(private ref: ElementRef, private renderer: Renderer2) {
		this.unsubscribe = renderer.listen(ref.nativeElement as HTMLElement, "click", (event: MouseEvent) => {
			if (this.skipNavigation) {
				event.stopImmediatePropagation();
			}
		});
	}

	public ngOnDestroy(): void {
		this.unsubscribe();
	}
}
```
